fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java

```diff
@@ -257,6 +261,11 @@ public <R> CompletableFuture<R> apply(Function<FDBRecordContext, CompletableFutu
                 lock.unlock(stamp);
                 commitIfNeeded();
                 return ret;
+            }).handle((ret, ex) -> {
+                if (ex != null) {
+                    reportAndReThrowException(ex);
+                }
+                return ret;
             });
```

Comment:
This changes the exception, wrapping it in an extra `CompletionException` (IIRC), it would be better to use [whenComplete](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletionStage.html#whenComplete(java.util.function.BiConsumer)) which doesn't change the exception or result at all.
When you do that you'll just want to log, but not `re-throw`

You could take a similar approach with `commit`, but you'd have to change it to `commitAsync`. That would be a good thing to do overall, but probably not something we want to bring into this PR, at least not in its entirety.

Reply:
✅